improv_ble_client/client.py

```diff
@@ -0,0 +1,521 @@
+            )
+            _LOGGER.debug("%s: Connected; RSSI: %s", self.name, self.rssi)
+
+            self._client = client
+
+            # Make sure the device has all improv characteristics
+            try:
+                self._resolve_characteristics(client.services)
+            except CharacteristicMissingError as err:
+                _LOGGER.debug(
+                    "%s: characteristic missing, clearing cache: %s; RSSI: %s",
+                    self.name,
+                    err,
+                    self.rssi,
+                    exc_info=True,
+                )
+                await client.clear_cache()
+                self._cancel_disconnect_timer()
+                await self._execute_disconnect_with_lock(DisconnectReason.ERROR)
+                raise
+
+            self._disconnect_reason = None
+            self._reset_disconnect_timer()
+
+            _LOGGER.debug(
+                "%s: Subscribe to notifications; RSSI: %s", self.name, self.rssi
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_ERROR, self._notification_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_RPC_RESULT, self._rpc_result_handler
+            )
+            await client.start_notify(
+                CHARACTERISTIC_UUID_STATE, self._notification_handler
+            )
+
+    def _resolve_characteristics(self, services: BleakGATTServiceCollection) -> None:
+        """Resolve characteristics."""
+        for characteristic in IMPROV_CHARACTERISTICS:
+            if not services.get_characteristic(characteristic):
+                raise CharacteristicMissingError(characteristic)
+
+    def _raise_if_not_connected(self) -> None:
+        """Raise if the connection to device is lost."""
+        if self._client and self._client.is_connected:
+            self._reset_disconnect_timer()
+            return
+        raise NotConnected
+
+    def _cancel_disconnect_timer(self):
+        """Cancel disconnect timer."""
+        if self._disconnect_timer:
+            self._disconnect_timer.cancel()
+            self._disconnect_timer = None
+
+    def _reset_disconnect_timer(self) -> None:
+        """Reset disconnect timer.
+
+        If the disconnect timer expires, disconnect from the device.
+        """
+
+        async def _disconnect() -> None:
+            """Execute disconnect request."""
+            _LOGGER.debug(
+                "%s: Disconnecting after timeout of %s",
+                self.name,
+                DISCONNECT_DELAY,
+            )
+            await self._execute_disconnect(DisconnectReason.TIMEOUT)
+
+        def _schedule_disconnect() -> None:
+            self._cancel_disconnect_timer()
+            disconnect_task = asyncio.create_task(_disconnect())
+            self._background_tasks.add(disconnect_task)
+            disconnect_task.add_done_callback(self._background_tasks.discard)
```

Comment:
This could be `self._async_create_background_task()` since you have the same pattern in `_disconnect`